### ISSUE TITLE
Used on table - display nested DataObjects

### ISSUE THREAD

**Issue by emteknetnz** [silverstripe/silverstripe-admin#1098]
### Overview
Currently the File Used on table presents duplicate information. If a file appears on a block, the used on table will have a record for the block, the page (and previously the elemental area). This duplicate information results in a rather confusing UI, where some rows contain links and others don't. These duplicates need to be reduced to one record.

To do this, we want to ensure there's enough context still given to what the object is and what the 'parent' object is. E.g for a file block, it would be useful to still see what 'parent' page it relates to.

### ACs
- For each use of a file only one record will show on the Used on table
- For each use of a file, show the object type it is being used on (e.g block, blog post, etc.) and a meaningful 'parent' object (e.g block page, blog, etc.)
- Ensure that files used multiple times within an object are still counted separately
- Each table row should include a link, directing the user to view the file in use (note: this may be further enhanced in: https://github.com/silverstripe/silverstripe-admin/issues/1100)
- There is a clear path to understand how links on the UsedOnTable to UserForm FileSubmissions can be added to the table (to then be tracked in https://github.com/silverstripe/silverstripe-userforms/issues/963)
- There is a clear path (including docs) for Developers to understand how to add their own data types to the table, if not added automatically
- This does not degrade performance from existing functionality where there are at least 20 relationships present in the table **Steve: OR we implement lazy loading on the used on table data**

### Designs
[Edit file / Used on tab](https://projects.invisionapp.com/dsm/silver-stripe/silverstripe-cms/nav/build/asset/componentContainers/5f152068f6772b33db2e52b0/tab/design?mode=preview)

### Early investigation notes
Estimated points: 8

Parent investigation issue https://github.com/silverstripeltd/product-issues/issues/249

Needs testing against kitchen sink to see how well it works in practice, more unit tests, possibly more extension PR's created for various data object types

#### PRs created as part of investigation (don't merge, though reuse bits)
- https://github.com/silverstripe/silverstripe-admin/pull/1097 (closed)

### Grooming notes
- It's likely that we'll run into performance issues due to the number of queries that could be required for each row
 - One idea to mitigate the performance issues could be to not load the table until the tab is selected to view (maybe this already happens?) **Steve: this sometimes happens, sometimes not, needs further investigation**
- There was a note about using pagination to support performance issues too

### PRs
- [x] https://github.com/silverstripe/silverstripe-framework/pull/9735
- [x] https://github.com/silverstripe/silverstripe-admin/pull/1139
- [x] https://github.com/silverstripe/silverstripe-asset-admin/pull/1139
- [x] https://github.com/silverstripe/silverstripe-versioned-admin/pull/174
- [x] https://github.com/dnadesign/silverstripe-elemental/pull/827
- [x] https://github.com/silverstripe/silverstripe-userhelp-content/pull/154 (docs)
- [x] https://github.com/silverstripe/silverstripe-framework/pull/9747 (docs)
- [x] https://github.com/silverstripe/silverstripe-userforms/pull/1015
- [x] https://github.com/silverstripe/silverstripe-elemental-fileblock/pull/18
- [x] https://github.com/silverstripe/silverstripe-elemental-bannerblock/pull/56
- [x] https://github.com/silverstripe/silverstripe-asset-admin/pull/1152 (fix)
- [x] https://github.com/silverstripe/silverstripe-framework/pull/9755/files (doc fix)

### Cucumber test
- https://studio.cucumber.io/projects/161118/test-plan/folders/1193246/scenarios/5486083

### Follow up issues
- https://github.com/silverstripe/silverstripe-asset-admin/issues/1146 - has been discussed and prioritised
- https://github.com/silverstripe/silverstripe-admin/issues/1100 - linking to elemental blocks
- https://github.com/silverstripe/silverstripe-admin/issues/1146 - linking to edit form tabs
- https://github.com/silverstripe/silverstripe-admin/issues/1147 - truncation

### Old PRs (closed)
- https://github.com/silverstripe/silverstripe-admin/pull/1121

### composer.json for testing

```json
    "require": {
        "php": ">=7.1.0",
        "silverstripe-themes/simple": "^3",
        "silverstripe/recipe-plugin": "^1",
        "silverstripe/recipe-cms": "4.x-dev",
        "silverstripe/recipe-content-blocks": "2.x-dev",
        "silverstripe/recipe-form-building": "1.x-dev",
        "dnadesign/silverstripe-elemental-userforms": "3.x-dev",

        "silverstripe/framework": "dev-pulls/4/dataobject-usage as 4.x-dev",
        "silverstripe/admin": "dev-pulls/1/usedontable-usage as 1.x-dev",
        "silverstripe/versioned-admin": "dev-pulls/1/usedontable-extension as 1.x-dev",
        "silverstripe/asset-admin":  "dev-pulls/1/usedontable-updates as 1.x-dev",
        "dnadesign/silverstripe-elemental": "dev-pulls/4/update-used-on-table-extension as 4.x-dev",
        "silverstripe/userforms": "dev-pulls/5/usedontable-extension as 5.x-dev",
        "silverstripe/elemental-bannerblock": "dev-pulls/2.1/singular-plural-names as 2.x-dev",
        "silverstripe/elemental-fileblock": "dev-pulls/2.1/singular-plural-names as 2.x-dev"

    },
    "repositories": [
        {
            "type": "vcs",
            "url": "git@github.com:creative-commoners/silverstripe-framework.git"
        },
        {
            "type": "vcs",
            "url": "git@github.com:creative-commoners/silverstripe-admin.git"
        },
        {
            "type": "vcs",
            "url": "git@github.com:creative-commoners/silverstripe-versioned-admin.git"
        },
        {
            "type": "vcs",
            "url": "git@github.com:creative-commoners/silverstripe-asset-admin.git"
        },
        {
            "type": "vcs",
            "url": "git@github.com:creative-commoners/silverstripe-elemental.git"
        },
        {
            "type": "vcs",
            "url": "git@github.com:creative-commoners/silverstripe-userforms.git"
        },
        {
            "type": "vcs",
            "url": "git@github.com:creative-commoners/silverstripe-elemental-bannerblock.git"
        },
        {
            "type": "vcs",
            "url": "git@github.com:creative-commoners/silverstripe-elemental-fileblock.git"
        }
    ],
```

**Comment by brynwhyman** [comment]
@emteknetnz I've updated the description to provide a bit more of an overview and some starting ACs

**Comment by brynwhyman** [comment]
Recent internal discussions noted that there's an emphasis on Developers using many-many-through objects for linking. These are probably objects that don't need to be included in the table, but being custom, we need a way to consider excluding them by default, or have that option be configurable.

@unclecheese mentioned that there should be a way to identify these linking objects at the outset.

**Comment by emteknetnz** [comment]
The used on table currently relies on the usages of `DataObject::$owns` to populate itself, via [RecursivePublishable::findOwners()](https://github.com/silverstripe/silverstripe-versioned/blob/1/src/RecursivePublishable.php#L197)

Files attached to a `DataObject::$has_one`, which is pretty common, DO NOT appear in the used on table.  This is a very significant short-coming

```php
<?php

namespace {

    use SilverStripe\AssetAdmin\Forms\UploadField;
    use SilverStripe\CMS\Model\SiteTree;
    use SilverStripe\Assets\File;

    class MyPage extends SiteTree
    {
        private static $db = [];

        private static $has_one = [
            'MyFile' => File::class
        ];

        public function getCMSFields()
        {
            $fields = parent::getCMSFields();
            $fields->insertBefore('Title', new UploadField('MyFile'));
            return $fields;
        }
    }
}
```

**Comment by emteknetnz** [comment]
Ditto for `DataObject::$many_many => ['MyFiles' => File::class]`

**Comment by michalkleiner** [comment]
We tried using the UsedOn table for pages, which our clients indicated would be very useful to see where a page has been linked from (via any ORM relationship — has_one, has_many or many_many) before they e.g. archive it or rename it etc. Having to add `$owns` everywhere (likely through extensions) to support it would be a massive job and probably not always desirable from ownership model perspective.

**Comment by chillu** [comment]
We could state that the reverse relationship needs to be established on `File` (via a `DataExtension`) for the records to show? Going through all DataObjects and all their relationships to identify which ones are files seems to compute intensive. That's more work than adding `$owns` though. The `lookupReverseOwners()` call is already uncached, so loops through *every* subclass of DataObject and checks `Config::inst()->get($class, 'owns')`. If we decide to infer that relationship without using `$owns` or `$belongs_to` etc, then I'd want to see some benchmarks on non-trivial projects (100+ DataObject types where a good subset has some kind of File relationship).

**Comment by emteknetnz** [comment]
@chillu roughly what I have so far is:

1) Loop everything in `Config::inst()->getAll()` which returns all types of DataObjects.  This is cached information so performance isn't a concern.
2) Go through the has_one, has_many, etc config values on each of these
3) See if the DataObject has a relationship to File, or a subclass of File
4) Do a raw sql query (or SQLSelect) that bypasses the ORM, so for `has_one` what I have so far is `SELECT ID FROM {$tableName} WHERE {$relationName}ID = {$recordID};`
  - $tableName will be something like `MyPage`
  - `$relationName` will be `MyFile`
  - `$recordID` will be the ID of the file you're looking for file usage info on

The main performance concern here will likely be the number of sql queries, the number of which will be in relation to the number of DataObjects that make use of Files.  Each query should be fairly quick as the WHERE clause will be on an indexed column.

**Comment by chillu** [comment]
I'd still want to see some performance comparisons from a non-trivial project (maybe ss.com?). Maybe ask the Terraformers on (roughly) how many queries this would produce on their big project? This logic is executed every single time a file is loaded for viewing or editing in AssetAdmin, as well as every single time you save it. And those are very common operations which just need to be fast. Alternatively, I think we've previously talked about lazy loading the "used on" table, there's no good reason it should hold up viewing or editing of files.

**Comment by emteknetnz** [comment]
>> I think we've previously talked about lazy loading the "used on" table, there's no good reason it should hold up viewing or editing of files.

Currently it's weirdly inconsistent when the logic for the used on table loads, sometimes it's only when you click on the used on table tab to explicitly load it, others it's whenever you select a different file.  I have not yet investigated what's causing this behaviour.

I think lazy-loading is the current priority to look at rather than performance of explicitly clicking something which is will be infrequently viewed.  I think it's if it's much better to have to wait a bit longer to get accurate results rather than what happens currently where only a subset of results is returned.

**Comment by chillu** [comment]
> I think it's if it's much better to have to wait a bit longer to get accurate results rather than what happens currently where only a subset of results is returned.

Yep agreed, particularly if it's only loaded once you open the tab. If it's loaded as a separate request in the background automatically after viewing/saving the record, it still increases webserver workloads unnecessarily, although it lowers the impact on the user (no delays in viewing/saving records)

**Comment by emteknetnz** [comment]
>> (no delays in viewing/saving records)

I'd say wait for real world experiences to see if this is an issue even worth solving before we dig to far into this.  I'm mindful of scope creep on what's already been estimated as a 13 point issue.

If this is perhaps something worth optimising, then we could make this a config value defaulted to off (better for webhosts) and let developers decide for themselves if they wish to turn it on for a better UX experience

**Comment by maxime-rainville** [comment]
I disagree with displaying all relations on the used on table. Everywhere else, we only look at "owned" relations:
- when building a change set
- when publishing a campaign or doing a recursive publish
- when giving a warning about deleting the file
- when restoring DataObjects

It's weird that this would be the one place where we disregard ownership. If we decide displaying all relations is the way to go, then we should rethink the behaviour in those other places and be systematic about it.

**Comment by emteknetnz** [comment]
Using ownership makes sense to me here:
- when building a change set
- when publishing a campaign or doing a recursive publish

Using ownership does not makes sense to me here:
- when giving a warning about deleting the file (very similar reasons as the used on table)

I'm not sure:
- when restoring DataObjects (I'm not sure what 'restoring' refers to)

**Comment by brynwhyman** [comment]
> I disagree with displaying all relations on the used on table.

When the ACs for this issue was crafted it wasn't clear that the logic had only ever been referencing 'owns' relationships. The original intention of this feature was really to focus on giving more context to the CMS user on _what object_ the image was being used by adding the breadcrumb and a (often more meaningful) parent object.

The fact that the used on table isn't accurately reflecting everywhere the file is being used seems like a flaw. The same can be said about the warning before deleting a file - the full impact of that action might not be accurately presented to the user. You're other examples sounds very specific to owned relationships though?

Apart from a mismatch between how many records a file deletion warning message quotes vs. the amount of records listed in the used on table, I don't see a reason why this shouldn't be implemented here (provided the lazy loading works for performance impacts)?.

**Comment by emteknetnz** [comment]
@brynwhyman @maxime-rainville 

I've listed the remaining discussion points.  I don't think anything here is blocking this being merged and beta tagged:

ManyManyThrough wasn't working
- https://github.com/silverstripe/silverstripe-framework/pull/9735#issuecomment-714791273
- I fixed this on Friday 23 Oct in a commit to framework

Logic used in "popup" when deleting a file
- https://github.com/silverstripe/silverstripe-framework/pull/9735#issuecomment-714791273
- Split off as a separate issue - https://github.com/silverstripe/silverstripe-asset-admin/issues/1146

Error messaging when table something goes wrong
- https://github.com/silverstripe/silverstripe-admin/pull/1139#issuecomment-714874275
- I said it is scope creep - we could create a new issue for this if we wanted to

Whether to display file relations (i.e. Pages, Content blocks) where CMS user does not have read permission to the relation.
- https://github.com/silverstripe/silverstripe-admin/pull/1139#discussion_r510566735
- This is easy to change later during beta. Probably worth have a wider discussion. I suggest we just keep as is for now

Move admin docs about making used on table extenions to docs.silverstripe.org
- https://github.com/silverstripe/silverstripe-admin/pull/1139/files#r509002754
- I'll do this later today

**Comment by brynwhyman** [comment]
Hey @emteknetnz I agree with your assessment of not letting these remaining issues block a merge. We're still waiting on a final UX review which I do think should hold up a merge until the findings from the review have been captured and reviewed. 

However...

> Whether to display file relations (i.e. Pages, Content blocks) where CMS user does not have read permission to the relation.

Agree that this shouldn't get an overhaul now, given you've outlined it's following a similar pattern to how the site tree behaves. Can we get it closer to the site tree behaviour now though? I.e Show a "Forbidden" toast notification instead of the page edit form. Then we're at a better place if we want to review this behaviour more broadly at a later time.

> Logic used in "popup" when deleting a file

FYI, I've pulled this into the Backlog. Would be good to keep this issue relevant.

**Comment by emteknetnz** [comment]
@brynwhyman The behaviour in the pull-request branch is consistent with the SiteTree behaviour - it will show a forbidden toast notification if a user without permission clicks on a page edit link.  There is no further modification of the pull-request required.

**Comment by brynwhyman** [comment]
Ah nice. In that case, that might not be an ideal end-state but IMO that's a separate concern from the work being done in this issue.

**Comment by maxime-rainville** [comment]
I'll have one final look ... but I'm happy to approve as-is.

On the "Error messaging when table something goes wrong", I'm not happy with JSON parsing errors being displayed, but at least the interface doesn't completely crash and remains mostly usable, so we can call this good enough. For the record, I don't think properly handling HTTP errors is ever scope creep ... whenever we do an HTTP request there's a non-trivial chance that it will fail and we should plan for those failures upfront.
